File: python/proofbundle/lineage.py
"""
Parent ref resolution, sealed_content_digest, verify_lineage
with cycle detection and depth budget.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Set, Callable

from .canonical import canonical_bytes
from .digest import digest


class LineageError(Exception):
    pass


def sealed_content_digest_input(bundle: Dict[str, Any]) -> Dict[str, Any]:
    """Build the input dict for sealed content digest computation."""
    result: Dict[str, Any] = {
        "hdr": bundle.get("hdr", {}),
        "payload": bundle.get("payload"),
    }
    meta = bundle.get("meta")
    if meta is not None:
        result["meta"] = meta
    refs = bundle.get("refs")
    if refs is not None:
        result["refs"] = refs
    return result


def sealed_content_digest(bundle: Dict[str, Any], alg: str) -> bytes:
    """Compute the sealed content digest of a bundle."""
    input_data = sealed_content_digest_input(bundle)
    data = canonical_bytes(input_data)
    return digest(alg, data)


def verify_lineage(
    bundle: Dict[str, Any],
    parent_resolver: Callable[[str, str], Optional[Dict[str, Any]]],
    max_depth: int = 256,
    visited: Optional[Set[str]] = None,
) -> Dict[str, Any]:
    """Verify the lineage chain of a bundle.
    
    Returns a dict with:
        - valid: bool
        - reason: str (optional)
        - depth: int
    """
    refs = bundle.get("refs", [])
    if not refs:
        return {"valid": True, "depth": 0}

    if visited is None:
        visited = set()

    if len(visited) >= max_depth:
        return {"valid": False, "reason": "max depth exceeded", "depth": len(visited)}

    bundle_id = bundle.get("hdr", {}).get("bundle_id", "")
    if bundle_id:
        if bundle_id in visited:
            return {"valid": False, "reason": "cycle detected", "depth": len(visited)}
        visited.add(bundle_id)

    for ref in refs:
        parent_id = ref.get("parent_id", "")
        parent_digest_b64u = ref.get("parent_digest", "")
        digest_alg = ref.get("digest_alg", "SHA-256")

        parent = parent_resolver(parent_id, parent_digest_b64u)
        if parent is None:
            return {"valid": False, "reason": f"parent not found: {parent_id}", "depth": len(visited)}

        # Verify parent digest
        computed = sealed_content_digest(parent, digest_alg)
        import base64
        computed_b64u = base64.urlsafe_b64encode(computed).rstrip(b"=").decode("ascii")
        if computed_b64u != parent_digest_b64u:
            return {"valid": False, "reason": f"parent digest mismatch: {parent_id}", "depth": len(visited)}

        # Recursively verify parent's lineage
        sub = verify_lineage(parent, parent_resolver, max_depth, set(visited))
        if not sub["valid"]:
            return {"valid": False, "reason": sub.get("reason"), "depth": sub.get("depth", len(visited))}

    return {"valid": True, "depth": len(visited)}

File: core/src/lineage.ts
/**
 * Parent reference resolution, sealedContentDigest, and lineage verification.
 * Includes cycle detection and depth budget enforcement.
 */

import { canonicalBytes } from './canonical.js';
import { digest } from './digest.js';
import type { Bundle, DigestAlgorithm } from './types.js';

export class LineageError extends Error {
  constructor(msg: string) {
    super(msg);
    this.name = 'LineageError';
  }
}

export interface LineageVerifyConfig {
  maxDepth: number;
  fetch: (id: string, digest: string) => Promise<Bundle | null>;
  trustedKeys?: Set<string>;
}

export interface LineageResult {
  valid: boolean;
  reason?: string;
  depth: number;
}

function stripSealFromHeader(hdr: Record<string, unknown>): Record<string, unknown> {
  const copy = { ...hdr };
  return copy;
}

function bytesToBase64(u8: Uint8Array): string {
  if (typeof globalThis !== 'undefined' && 'Buffer' in globalThis) {
    return (globalThis as unknown as { Buffer: { from: (u: Uint8Array) => { toString: (enc: string) => string } } }).Buffer.from(u8).toString('base64');
  }
  const bin = Array.from(u8, b => String.fromCharCode(b)).join('');
  return btoa(bin);
}

export function sealedContentDigestInput(bundle: Bundle): Record<string, unknown> {
  const result: Record<string, unknown> = {
    hdr: stripSealFromHeader(bundle.hdr as Record<string, unknown>),
    payload: bundle.payload,
  };
  if (bundle.meta !== undefined) {
    result.meta = bundle.meta as Record<string, unknown>;
  }
  if (bundle.refs !== undefined && bundle.refs.length > 0) {
    result.refs = bundle.refs;
  }
  return result;
}

export async function sealedContentDigest(
  bundle: Bundle,
  alg: DigestAlgorithm
): Promise<Uint8Array> {
  const input = sealedContentDigestInput(bundle);
  const bytes = canonicalBytes(input);
  return digest(alg, bytes);
}

export async function verifyLineage(
  bundle: Bundle,
  config: LineageVerifyConfig,
  visited: Set<string> = new Set()
): Promise<LineageResult> {
  if (!bundle.refs || bundle.refs.length === 0) {
    return { valid: true, depth: 0 };
  }

  if (visited.size >= config.maxDepth) {
    return { valid: false, reason: 'resource-exhausted: max depth exceeded', depth: visited.size };
  }

  const hdrId = bundle.hdr.bundle_id ?? '';
  if (hdrId) {
    if (visited.has(hdrId)) {
      return { valid: false, reason: 'cycle detected', depth: visited.size };
    }
    visited.add(hdrId);
  }

  for (const ref of bundle.refs) {
    const parent = await config.fetch(ref.id, ref.digest);
    if (!parent) {
      return { valid: false, reason: `parent not found: ${ref.id}`, depth: visited.size };
    }

    const parentDigest = await sealedContentDigest(parent, ref.digest_alg);
    const parentDigestB64 = bytesToBase64(parentDigest);
    if (parentDigestB64 !== ref.digest) {
      return { valid: false, reason: `parent digest mismatch: ${ref.id}`, depth: visited.size };
    }

    if (config.trustedKeys && config.trustedKeys.size > 0) {
      const pkMatch = config.trustedKeys.has(ref.public_key);
      if (!pkMatch) {
        return { valid: false, reason: `parent key not trusted: ${ref.id}`, depth: visited.size };
      }
    }

    const sub = await verifyLineage(parent, config, new Set(visited));
    if (!sub.valid) {
      return { valid: false, reason: sub.reason, depth: sub.depth };
    }
  }

  return { valid: true, depth: visited.size };
}
